File: app/src/main/java/com/tunjid/rcswitchcontrol/adapters/RemoteSwitchAdapter.java
package com.tunjid.rcswitchcontrol.adapters;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Switch;
import android.widget.TextView;

import com.tunjid.rcswitchcontrol.R;
import com.tunjid.rcswitchcontrol.model.RcSwitch;

import java.util.List;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;


public class RemoteSwitchAdapter extends RecyclerView.Adapter<RemoteSwitchAdapter.ViewHolder> {

    private static final int BLE_DEVICE = 1;

    private List<RcSwitch> switches;
    private SwitchListener switchListener;

    public RemoteSwitchAdapter(SwitchListener switchListener, List<RcSwitch> switches) {
        this.switchListener = switchListener;
        this.switches = switches;
    }

    @NonNull @Override
    public ViewHolder onCreateViewHolder(@NonNull ViewGroup viewGroup, int viewType) {
        Context context = viewGroup.getContext();
        View itemView = LayoutInflater.from(context).inflate(R.layout.viewholder_remote_switch, viewGroup, false);

        return new ViewHolder(itemView);
    }

    @Override
    public void onBindViewHolder(@NonNull ViewHolder viewHolder, final int position) {
        viewHolder.bind(switches.get(position), switchListener);
    }

    @Override
    public int getItemViewType(int position) {
        return BLE_DEVICE;
    }

    @Override
    public int getItemCount() {
        return switches.size();
    }

    // ViewHolder for actual content
    public static class ViewHolder extends RecyclerView.ViewHolder {

        TextView deviceName;
        Switch toggle;

        RcSwitch rcSwitch;
        SwitchListener switchListener;

        ViewHolder(View itemView) {
            super(itemView);
            deviceName = itemView.findViewById(R.id.switch_name);
            toggle = itemView.findViewById(R.id.switch_toggle);
        }

        void bind(RcSwitch rcSwitch, SwitchListener switchListener) {
            this.rcSwitch = rcSwitch;
            this.switchListener = switchListener;

            deviceName.setText(rcSwitch.getName());

            toggle.setOnClickListener(view -> switchListener.onSwitchToggled(rcSwitch, toggle.isChecked()));
            itemView.setOnLongClickListener(view -> {
                switchListener.onLongClicked(rcSwitch);
                return true;
            });
        }
    }

    public interface SwitchListener {
        void onLongClicked(RcSwitch rcSwitch);

        void onSwitchToggled(RcSwitch rcSwitch, boolean state);
    }

}

File: app/src/main/java/com/tunjid/rcswitchcontrol/adapters/RemoteSwitchAdapter.kt
package com.tunjid.rcswitchcontrol.adapters

import android.view.ViewGroup
import com.tunjid.androidbootstrap.recyclerview.InteractiveAdapter
import com.tunjid.rcswitchcontrol.R
import com.tunjid.rcswitchcontrol.data.Device
import com.tunjid.rcswitchcontrol.data.RcSwitch
import com.tunjid.rcswitchcontrol.data.ZigBeeDevice

typealias ViewHolder =  DeviceViewHolder<out InteractiveAdapter.AdapterListener, out Device>

class RemoteSwitchAdapter(
        switchListener: Listener,
        private val switches: List<Device>
) : InteractiveAdapter<ViewHolder, RemoteSwitchAdapter.Listener>(switchListener) {

    init {
        setHasStableIds(true)
    }

    override fun onCreateViewHolder(viewGroup: ViewGroup, viewType: Int): ViewHolder = when (viewType) {
        RF_DEVICE -> RfDeviceViewHolder(getItemView(R.layout.viewholder_remote_switch, viewGroup), adapterListener)
        ZIGBEE_DEVICE -> ZigBeeDeviceViewHolder(getItemView(R.layout.viewholder_remote_switch, viewGroup), adapterListener)
        else -> throw IllegalArgumentException("Invalid object type")
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) = when (val device = switches[position]) {
        is RcSwitch -> (holder as RfDeviceViewHolder).bind(device)
        is ZigBeeDevice -> (holder as ZigBeeDeviceViewHolder).bind(device)
        else -> throw IllegalArgumentException("Invalid object type")
    }

    override fun getItemViewType(position: Int): Int = when (switches[position]) {
        is RcSwitch -> RF_DEVICE
        is ZigBeeDevice -> ZIGBEE_DEVICE
        else -> throw IllegalArgumentException("Invalid object type")
    }

    override fun getItemCount(): Int = switches.size

    override fun getItemId(position: Int): Long = switches[position].hashCode().toLong()

    interface Listener : AdapterListener, RfDeviceViewHolder.Listener, ZigBeeDeviceViewHolder.Listener

    companion object {

        private const val RF_DEVICE = 1
        private const val ZIGBEE_DEVICE = 2

    }

}
